### PR TITLE
Fix validation of rollback nodes

### PR DESCRIPTION
**Pull request by cocreature**
To my own embarassment there is no test in this PR. We don’t have any
direct tests for validation atm. We absolutely need some and I already
discussed with Remy that this is something we should focus on once
exceptions are done but for now I’ll stick to being sad.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala [hunk 1]
```diff
@@ -288,6 +288,21 @@ private final class Validation[Nid, Cid](implicit ECid: Equal[Cid]) {
...
+                  Node.NodeRollback(
+                    children2,
+                    version2,
+                  ),
+                ) if version1 == version2 =>
+              loop(
+                children1.iterator.to(LazyList),
+                children2.iterator.to(LazyList),
+                stack,
+              )
```
Comment (nickchapman-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala:305]: We have dropped `rest1` and `rest2` on the floor.
Don't we need to still process these?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala [hunk 2]
```diff
@@ -288,6 +288,21 @@ private final class Validation[Nid, Cid](implicit ECid: Equal[Cid]) {
...
             case _ =>
               Left(ReplayNodeMismatch(recorded, nid1, replayed, nid2))
```
Comment (nickchapman-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala:307]: I'm sad the scala compiler didn't discover statically that the case of `Rollback` nodes was missing. Initially I though the blame was our this catch-all...
```
            case _ =>
              Left(ReplayNodeMismatch(recorded, nid1, replayed, nid2))
```

And I wondered if being more precise would have provoked the desired unmatched case warnings when the new node type was added....

But no. There are no warnings even if the catch-all is removed. Scala is broken  :(

**Review by nickchapman-da — CHANGES_REQUESTED** [review]
I think there is a bug,

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala (hunk 1 above)
Comment (cocreature) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala:318]: good point, I’ve pushed them to the stack like we do for exercise nodes 912b0f9e8e90b6f26ecfcc42455a1d5977d36016

Comment (nickchapman-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala:318]: :+1:

**Review by nickchapman-da — APPROVED** [review]

**Pull request merged**